README.md
# TWOMU

This War of Mine Utils

# What is This?

This app makes it possible to use the W/A/S/D keys to control the camera in This War of Mine, <br>as well as adding several features, like hiding character outlines, or the pencil effect from the game.

<figure>
    <img src="demo.gif">
    <figcaption>(old) Video Demo</figcaption>
</figure>

# How does this work?

It's pretty simple, when the app detects that the W/A/S/D key(s) are held down, <br>or when you enabled a function from the terminal, it writes memory to specific addresses.

# How can I use it?

Download the app from the [Releases page](https://github.com/cmd777/TWOMBC/releases/latest), then run it. simple as that.

Step -> determines how fast/much the camera should move when pressing W/A/S/D, The default value is 0.7

ReadMem -> will periodically (every 10ms) write the X, Y coordinates from the game's memory to a stored one<br>This fixes an issue where pressing tab or using the mouse to change camera position would rubberband the camera back.

FixCam -> will periodically (every 10ms) check, and set a value to an address that controls the camera mode.<br>This fixes a notorious issue, that when you loaded into a level, or moved the camera by other means, would disable the ability to use W/A/S/D controls.

DisablePencil -> will disable the in-game pencil effect.<br>Note: this doesn't have an effect on frame rate.

DisableRain -> will disable the in-game rain effect.<br>Note: this doesn't have an effect on frame rate.

DisableOutlines -> will disable the flashing character outlines.<br>Note: this doesn't have an effect on frame rate.

ModifyWndProc -> will change (NOP) TWOM's WndProc WM_SIZE.<br>Whenever TWOM is minimized, and then reopened, there is about a 2s black screen before the game can show anything (1s because of kernel32's 1000ms sleep, and another one rendering everything), ModifyWndProc will NOP the 'if' condition to WM_SIZE, and make the 2s process near instantaneous<br>This comes at a downside, as attempting to resize the game from anything lower than 100% resolution back to 100% makes everything low resolution (NOTE: It's possible to change resolution back to 100%, it needs to be done from the settings menu.) other than that, there are no other downsides found.

Randomize Settings, Randomize Locations -> will randomly set values in the `My Story` tab<br>
Randomize Settings requires the `My Story` tab to be open <br>
Randomize Locations requires the `My City Map` tab to be open.

# Nothing is happening | Weird characters in the terminal
If you haven't, when TWOM is started, make sure to run option 1 in TWOMU. This will find the game's PID and Handle.

If you have any option enabled, but is not working, run [DebugView](https://learn.microsoft.com/en-us/sysinternals/downloads/debugview), and check the logs.
<br>If you believe this is an issue with TWOMU, submit a bug report.

If you see any unusual characters like &larr;, that is because your terminal doesn't support ANSI escape sequences.
If you are using any windows version after 1511 (build number 10586), then you can easily fix this.
<br>Open regedit, and go to `HKEY_CURRENT_USER\Console`, then create a DWORD (32-bit value) with the name `VirtualTerminalLevel`, and set the value to 1

TBD: SetConsoleTextAttribute, that should work on most windows versions.

# Known Issues
Terminals that don't support ANSI escape sequences have &larr; & no color.<br>
PReadMem breaks or bugs out Stories' intros & animations.<br>
Attempting to set random values in the My Stories tab after leaving a previously saved game will break it.<br>

# My antivirus says it's a virus, is it?

<figure>
    <blockquote>
        <p>
            Commercial virus scanning programs are often confused by the structure of Go binaries, <br> which they don't see as often as those compiled from other languages.
        </p>
    </blockquote>
    <figcaption>Excerpt from <a href="https://go.dev/doc/faq#virus">https://go.dev/doc/faq#virus</a></figcaption>
</figure>

Aside from the antivirus being confused by the app's structure, it may also detect that the program imports user32.dll, kernel32.dll, psapi.dll, and that it calls function like GetAsyncKeyState, Read/Write ProcessMemory, it will mark the app as a virus.

There is an extremely high chance that the antivirus will detect TWOMUHook.dll as well.

**Why?**

It does the same things the Go version does, except, it hooks onto DirectX's EndScene using Detours, and that's a big no-no from antiviruses.

It's open source, you can build everything for yourself, and change things that don't suit your liking

# Can I run it on Linux?

Yes, you can! It requires some tinkering though.

1. Download the latest [TWOMU](https://github.com/cmd777/TWOMBC/releases/latest) Version
2. Install [Wine](https://wiki.winehq.org/Download)
3. Change Steam Compatability to Proton
    - Right click on This War of Mine
    - Go to Properties
    - Click on Compatability
    - Check `Force the use of a specific Steam Play compatability tool`
    - Select the latest proton version (right under Proton Experimental)
4. Paste the TWOMUHook directory to `/<STEAM_PATH>/SteamLibrary/steamapps/compatdata/282070/pfx` (If you wish to use the ImGui version, if not, skip this step.)
5. Open your terminal, and paste the following command
    - `STEAM_COMPAT_CLIENT_INSTALL_PATH=... STEAM_COMPAT_DATA_PATH=... WINEPREFIX=$PWD {PROTONPATH} run "/<TWOMU_INSTALL_PATH>/TWOMU-x64.exe"`

STEAM_COMPAT_CLIENT_INSTALL_PATH is where ".steam/steam" is. (if `apt install steam` was used, it should be `"$HOME/.steam/steam"`)

STEAM_COMPAT_DATA_PATH is where steam installed the game (for example, `"/<STEAM_PATH>/SteamLibrary/steamapps/compatdata/282070"`)

PROTONPATH is where proton is installed (should be in steam's root directory, `"/home/<USERNAME>/.steam/root/steamapps/common/Proton 7.0/proton"`)

## Limitations

Unforunately, it's not perfect yet, and some functions don't work well.

| Function                  | Compatability         |
| :---:                     | :---:                 |
| UseWASD                   | :heavy_check_mark:    |
| ReadMemory                | :heavy_check_mark:    |
| FixCamera                 | :heavy_check_mark:    |
| Disable Pencil Effect     | :interrobang:         |
| Disable Rain Effect       | :interrobang:         |
| Disable Outlines          | :interrobang:         |
| Modify WndProc            | :heavy_check_mark:    |

:heavy_check_mark: - The function works without issues

:interrobang: - The function works, but is prone to bugs.


CHANGELOG.md
## [v3.0.0](https://github.com/cmd777/TWOMU/releases/tag/v3.0.0) (2022-12-04)

### Added
- MSBuild to build TWOMUHook
- Solution files, they can be found in the `TWOMUHook` directory (compiling only works with the x64 version)
- Regex to Go, this will find both cls and clear without case sensitivity
- New option to Inject TWOMUHook (q)
- Disable Pencil Effect, Disable Rain Effect, Disable Character Outlines, Modify WndProc to the ImGui version
- NOP, Offsets Function to the ImGui version
- WASD Controls to ImGui
- Keyboard navigation option to ImGui

### Changed
- (Workflows) Runner to Windows
- Files are now zipped
- All inputs in Go are now lowercase, that means that Q will be registered as q

<hr>

## [v3.0.0-pre.4](https://github.com/cmd777/TWOMU/releases/tag/v3.0.0-pre.4) (2022-12-02)
## Added
- NOP, Offsets Function to the ImGui version
- WASD Controls to ImGui
- Keyboard navigation option to ImGui
## Removed
- Nop Arrays from the ImGui version

<hr>

## [v3.0.0-pre.3](https://github.com/cmd777/TWOMU/releases/tag/v3.0.0-pre.3) (2022-11-29)
## Added
- Regex to Go, this will find both cls and clear without case sensitivity
- New option to Inject TWOMUHook (q)
- Disable Pencil Effect, Disable Rain Effect, Disable Character Outlines, Modify WndProc to the ImGui version
## Changed
- All inputs in Go are now lowercase, that means that Q will be registered as q

<hr>

## [v3.0.0-pre.2](https://github.com/cmd777/TWOMU/releases/tag/v3.0.0-pre.2) (2022-11-28)
## Added
- MSBuild to build TWOMUHook
- The built dll file to the zip file (the option to inject it is the same as the option used to find This War of Mine and calculate offsets - This will be changed.)
- Partially implemented ImGui (only the demo window is shown)
- Solution files, they can be found in the `TWOMUHook` directory (compiling only works with the x64 version)

<hr>

## [v3.0.0-pre.1](https://github.com/cmd777/TWOMU/releases/tag/v3.0.0-pre.1) (2022-11-25)

### Added
- (Workflows) G++ to build DllMain
- Custom "hasher" function to write CHANGELOG to Releases

### Changed
- (Workflows) Runner to Windows
- Files are now zipped

<hr>

## [v2.0.1](https://github.com/cmd777/TWOMU/releases/tag/v2.0.1) (2022-11-25)

### Added
- A patch option to allow multiple running instances of This War of Mine


### Removed
- Redundant forever loop in PrintPatches


<hr>

## [v2.0.1-pre.2](https://github.com/cmd777/TWOMU/releases/tag/v2.0.1-pre.2) (2022-11-23)
### Removed
- Redundant forever loop in PrintPatches

## [v2.0.1-pre.1](https://github.com/cmd777/TWOMU/releases/tag/v2.0.1-pre.1) (2022-11-23)
### Added
- A patch option to allow multiple running instances of This War of Mine

<hr>

## [v2.0.0](https://github.com/cmd777/TWOMU/releases/tag/v2.0.0) (2022-11-17)

### Added
- An option to randomize the My Stories & My City Map
- DisableOutlines, which disable the constantly flashing character outlines
- Terminal now has color, this requires a windows version after 1511 (build number 10586) and VirtualTerminalLevel to be set
### Changed
- Name from TWOMBC to TWOMU
- Errors will now be printed to Debug Output, and not the console
- Values can now be changed back from within the console
- Foreground application checking was merged into the WASD controls

### Removed
- Flags, and added a semi-interactive "interface"

<hr>

## [v1.0.4](https://github.com/cmd777/TWOMU/releases/tag/v1.0.4) (2022-11-05)

### Added
- ModifyWndProc, which if set to true, will change (NOP) TWOM's WndProc WM_SIZE

<hr>

## [v1.0.3](https://github.com/cmd777/TWOMU/releases/tag/v1.0.3) (2022-11-04)

### Added
- DisablePencil, which if set to true, will disable the in-game pencil effect (Note: this doesn't have an effect on frame rate)
- DisableRain, which if set to true, will disable the in-game rain effect (Note: this doesn't have an effect on frame rate)

<hr>

## [v1.0.2](https://github.com/cmd777/TWOMU/releases/tag/v1.0.2) (2022-11-02)

### Added
- PrintErr, which if set to true, will print any errors that show up (default value is false, as it can get quite spammy)
- Functions for CheckFG, ReadMem, FixCam
### Changed
- Step's default value from 0.3 -> 0.7

### Fixed
- A lot of typos, and updated the README
- Potential deadlock to time.Sleep with time.Timer

<hr>

## [v1.0.1](https://github.com/cmd777/TWOMU/releases/tag/v1.0.1) (2022-11-01)

### Added
- PReadMem, which will periodically (every 10ms) write the X, Y coordinates from the game's memory to a stored one
- Mutexes, they are locked at the beginning of a goroutine, and unlocked at the end (This fixes a data race)

### Changed
- Step's default value from 0.2 -> 0.3

hasher/hash.go
package main

import (
	"crypto/sha256"
	"fmt"
	"io"
	"os"
	"strings"
)

func main() {
	var pretext string

	chlog_body, err := os.ReadFile("../CHANGELOG.md")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	if strings.Contains(os.Args[1], "pre") {
		pretext = "<p align='center'><img src='https://img.shields.io/badge/-This%20is%20an%20experimental%20build%20and%20may%20not%20be%20fully%20stable-orange?style=plastic'></p>"
	}

	file, err := os.Open(os.Args[2])
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	defer file.Close()

	hash := sha256.New()
	if _, err := io.Copy(hash, file); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	fulltext := fmt.Sprintf("%v\n\n%v\n\n%v SHA256: %x", pretext, string(chlog_body), strings.Replace(os.Args[2], "../", "", -1), hash.Sum(nil))

	errw := os.WriteFile("../CHANGELOG.md", []byte(fulltext), 0666)
	if errw != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	fmt.Println("Ok")
}


TWOMUHook/twomu_hooks.h
#pragma once

#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <Psapi.h>
#include <string>
#include <mutex>
#include <dwmapi.h>

#include "dep/dxsdk/d3d9.h"
#include "dep/dxsdk/d3dx9.h"

#include "dep/detours/detours.h"

#include "dep/imgui/imgui.h"
#include "dep/imgui/imgui_impl_dx9.h"
#include "dep/imgui/imgui_impl_win32.h"

#pragma comment (lib, "dep/dxsdk/d3d9.lib")
#pragma comment (lib, "dep/dxsdk/d3dx9.lib")
#pragma comment (lib, "dep/detours/detours.lib")

#define TWOMU_VERSION "v3.0.1-EXPERIMENTAL"

WNDPROC V_WNDPROC = NULL;
HWND V_HWND = NULL;
HMODULE V_HMODULE = NULL;

typedef HRESULT(__stdcall* EndScene)(IDirect3DDevice9* d3ddev9);
EndScene V_EndScene;

typedef HRESULT(__stdcall* Reset)(IDirect3DDevice9* d3ddev9, D3DPRESENT_PARAMETERS* d3dpp);
Reset V_Reset;

bool SHOW = true;
bool INIT = false;

bool UseWASD = false;
bool ReadMemory = false;
bool FixCamera = false;
bool DisablePEffect = false;
bool DisableRainEffect = false;
bool DisableOutlines = false;
bool ModifyWndProc = false;

DWORD PID;
HANDLE TWOMHandle;
INT64 BaseAddr;

byte V_Pencil[] = { 0xF3, 0x44, 0x0F, 0x10, 0x05, 0x35, 0x10, 0x4A, 0x00 };
byte V_Rain[] = { 0x0F, 0xB7, 0x05, 0xB8, 0xD7, 0x72, 0x00 };
byte V_Outlines[] = { 0xF3, 0x0F, 0x10, 0x0D, 0xD8, 0x2F, 0x60, 0x00 };
byte V_MWndProc[] = { 0x83, 0xE8, 0x02, 0x74, 0x26 };

INT64 Pencil;
INT64 Rain;
INT64 Outlines;
INT64 MWndProc;

INT64 XArray[] = { 0x70 };
INT64 YArray[] = { 0x78 };
INT64 CArray[] = { 0xA6 };
INT64 XPos;
INT64 YPos;
INT64 CMode;

float Step = 0.2f;
float X = 0.0f;
float Y = 0.0f;

int CM;
int IdealCameraMode = 148602;
int DWMDarkThemeValue = 1;

std::mutex Mutex;

extern IMGUI_IMPL_API LRESULT ImGui_ImplWin32_WndProcHandler(HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam);

main.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	memory "main/memory"
	"math"
	"math/rand"
	"os"
	"regexp"
	"strings"
	"sync"
	"syscall"
	"time"
	"unsafe"
)

const (
	Abandoned_Cottage = iota
	Shelled_Cottage
	Ruined_Block_of_Flats
	Looted_Gas_Station
	Ghost_House
	Garage
	Small_Apartment_Building
	Decrepit_Squad
	Brothel
	Shelled_School
	Warehouse
	Ruined_Villa
	Semi_Detached_House
	Military_Outpost
	Hotel
	Construction_Site
	Quiet_House
	Supermarket
	Sniper_Junction
	St_Marys_Church
	City_Hospital
	Old_Town
	Port
	Airport
	Pharmacy
	Ruined_Toy_Store
	Park
	Bakery
	Shelled_Brewery
	The_Samuel_Institute
	Railway_Station
	Music_Club
)

var (
	Step float32 = 0.7

	UseWASD           bool = false
	PReadMem          bool = false
	PFixCam           bool = false
	DisablePEffect    bool = false
	DisableRainEffect bool = false
	DisableOutlines   bool = false
	ModifyWndProc     bool = false

	WASDChan    = make(chan struct{})
	ReadMemChan = make(chan struct{})
	FixCamChan  = make(chan struct{})

	TWOMPID  uint32
	BaseAddr int64

	HANDLE, XPos, YPos, CMode, Rain, Pencil, WndProc, Outline uintptr
	XBuffer, YBuffer, CMBuffer                                []byte
	X, Y                                                      float32
	CM                                                        uint32

	// For Custom Scenarios
	LocationsPTR      uintptr
	LocationsAddr     []uintptr
	LocationsToChoose int
	LocationsAdded    = make(map[int]uintptr)
	LocationsData     = map[int]int{
		Abandoned_Cottage:        1,
		Shelled_Cottage:          1,
		Ruined_Block_of_Flats:    1,
		Looted_Gas_Station:       1,
		Ghost_House:              1,
		Garage:                   1,
		Small_Apartment_Building: 2,
		Decrepit_Squad:           1,
		Brothel:                  1,
		Shelled_School:           2,
		Warehouse:                1,
		Ruined_Villa:             2,
		Semi_Detached_House:      2,
		Military_Outpost:         1,
		Hotel:                    3,
		Construction_Site:        2,
		Quiet_House:              1,
		Supermarket:              2,
		Sniper_Junction:          2,
		St_Marys_Church:          2,
		City_Hospital:            1,
		Old_Town:                 1,
		Port:                     1,
		Airport:                  1,
		Pharmacy:                 1,
		Ruined_Toy_Store:         1,
		Park:                     1,
		Bakery:                   1,
		Shelled_Brewery:          1,
		The_Samuel_Institute:     1,
		Railway_Station:          1,
		Music_Club:               2,
	}

	Ceasefirep, Intensityp, WinterComesp, WinterHarshnessp, WinterLengthp uintptr
	Ceasefire, Intensity, WinterComes, WinterHarshness, WinterLength      int

	ExitRegex  = regexp.MustCompile(`(?mi)(exit|quit)`)
	ClearRegex = regexp.MustCompile(`(?mi)(clear|cls)`)

	Mutex sync.Mutex

	S = time.NewTimer(10 * time.Millisecond)
)

func GetTWOM() {
	var entry syscall.ProcessEntry32
	entry.Size = uint32(unsafe.Sizeof(syscall.ProcessEntry32{}))

	snhandle := memory.CreateToolhelp32Snapshot()

	ok, snapshot := memory.Process32FirstW(snhandle, entry)
	if ok {
		for {
			okproc, process := memory.Process32NextW(snhandle, snapshot)
			if !okproc {
				break
			}

			if strings.Contains(syscall.UTF16ToString(process.ExeFile[:]), "This War of Mine.exe") {
				TWOMPID = process.ProcessID
				break
			}
		}
	}

	HANDLE = memory.OpenProcess(TWOMPID)

	Modules := memory.EnumProcessModules(HANDLE)

	for i := 0; i < len(Modules); i++ {
		ModuleFileName := memory.GetModuleFileNameExW(HANDLE, Modules[i])
		if strings.Contains(ModuleFileName, "This War of Mine.exe") {
			ModuleInfo := memory.GetModuleInformation(HANDLE, Modules[i])
			BaseAddr = int64(ModuleInfo.LpBaseOfDll)
			break
		}
	}

	XPos = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x009064D0, 0x70)
	XBuffer = memory.ReadProcessMemory(HANDLE, XPos, 8)
	X = math.Float32frombits(binary.LittleEndian.Uint32(XBuffer))

	YPos = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x009064D0, 0x78)
	YBuffer = memory.ReadProcessMemory(HANDLE, YPos, 8)
	Y = math.Float32frombits(binary.LittleEndian.Uint32(YBuffer))

	CMode = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x009064D0, 0xA6)

	// f3 44 0f 10 05 35 10 4a 00
	// MOVSS XMM8,[TWOM+6EE7C0]
	Pencil = uintptr(BaseAddr) + 0x24D782

	// 0f b7 05 b8 d7 72 00
	// MOVZX EAX,word ptr [TWOM+8E1ADE]
	Rain = uintptr(BaseAddr) + 0x1B431F

	//	83 e8 02
	//	SUB EAX, 0x2
	//	74 26
	//	JZ TWOM+4C2C5C
	WndProc = uintptr(BaseAddr) + 0x4C2C31

	//	f3 0f 10 0d d8 2f 60 00
	//	MOVSS XMM1, dword ptr [TWOM+819490]
	Outline = uintptr(BaseAddr) + 0x2164B0

	PrintMenu()
}

func PrintPatches() {
	fmt.Println()
	fmt.Println("This War of Mine Utils -> Patches")
	fmt.Println("The following options will patch the game executable itself.")
	fmt.Println("Whilst it shouldn't happen, make sure you have a backup of your savefiles in case of a corruption caused by the patched exe.")
	fmt.Println()

	// pretty pointless for now.. maybe i'll do something with it later
	fmt.Println("allow-more-instances) Will patch the game to allow more than 1 running instance")
	fmt.Println("cls|clear) Clear Screen")
	fmt.Println("leave) Go Back to The Normal Menu")

	var Option string

	fmt.Scan(&Option)

	switch Option {
	case ClearRegex.FindString(Option):
		ClearScreen()
		PrintPatches()
	case "allow-more-instances":
		var ExePath string

		fmt.Print("Please specify where This War of Mine.exe is (or just drag the exe here) -> ")

		inreader := bufio.NewScanner(os.Stdin)

		for {
			inreader.Scan()
			if ExePath = inreader.Text(); ExePath != "" {
				break
			}
		}

		ExePath = strings.Replace(ExePath, "\"", "", -1)

		twombytes, err := os.ReadFile(ExePath)
		if err != nil {
			fmt.Println("Failed to Read File -> ", err)
		}
		fmt.Println("File Read OK")

		Addr := memory.ScanBytes(twombytes, []byte{0x48, 0x85, 0xC0, 0x74, 0x3A, 0x48, 0x8B, 0xC8, 0xFF, 0x15, 0x08, 0x33, 0x0B, 0x00})
		if Addr == 0 {
			fmt.Println("Failed to Find Bytes")
		} else {
			twombytes[Addr+3] = 0xEB
		}

		errw := os.WriteFile("twom-mi.exe", twombytes, 0666)
		if errw != nil {
			fmt.Println("Failed to Write File -> ", errw)
		}

		if err == nil && errw == nil {
			fmt.Println("twom-mi.exe was written successfully. leaving patches menu.")
		} else {
			fmt.Println("Something failed while attempting to patch the exe, leaving patches menu.")
		}
	case "leave":
		ClearScreen()
		PrintMenu()
	default:
		fmt.Println(Option, "is not a valid option.")
	}
}

func PrintMenu() {
	fmt.Println()
	fmt.Println("This War of Mine Utils")
	fmt.Println("1) Find Handle of TWOM and calculate offsets")

	if UseWASD {
		fmt.Println("\033[92m[on]\033[0m\t2) Use WASD Controls")
	} else {
		fmt.Println("\033[91m[off]\033[0m\t2) Use WASD Controls")
	}

	if PReadMem {
		fmt.Println("\033[92m[on]\033[0m\t3) Read & Write game memory to a stored one")
	} else {
		fmt.Println("\033[91m[off]\033[0m\t3) Read & Write game memory to a stored one")
	}

	if PFixCam {
		fmt.Println("\033[92m[on]\033[0m\t4) Fix Camera for WASD Controls")
	} else {
		fmt.Println("\033[91m[off]\033[0m\t4) Fix Camera for WASD Controls")
	}

	if DisablePEffect {
		fmt.Println("\033[92m[on]\033[0m\t5) Hide Pencil Effect")
	} else {
		fmt.Println("\033[91m[off]\033[0m\t5) Hide Pencil Effect")
	}

	if DisableRainEffect {
		fmt.Println("\033[92m[on]\033[0m\t6) Hide Rain")
	} else {
		fmt.Println("\033[91m[off]\033[0m\t6) Hide Rain")
	}

	if DisableOutlines {
		fmt.Println("\033[92m[on]\033[0m\t7) Hide Outlines")
	} else {
		fmt.Println("\033[91m[off]\033[0m\t7) Hide Outlines")
	}

	if ModifyWndProc {
		fmt.Println("\033[92m[on]\033[0m\t8) Modify WndProc")
	} else {
		fmt.Println("\033[91m[off]\033[0m\t8) Modify WndProc")
	}

	fmt.Println()
	fmt.Println("----------------------------------------")
	fmt.Println()

	fmt.Println("i) Randomize Settings\t(Custom Scenario)")
	fmt.Println("o) Randomize Locations\t(Custom Scenario)")

	fmt.Println("0) Change Step Value ( now", Step, ")")

	fmt.Println("a) Toggle All")
	fmt.Println("p) Patches")
	fmt.Println("q) Inject TWOMUHook")
	fmt.Println("cls|clear) Clear Screen")
	fmt.Println("exit|quit Exit TWOMU")
}

func ClearScreen() {
	fmt.Print("\033[2J")
	fmt.Print("\033[H")
}

func main() {
	var Option string

	PrintMenu()

	for {
		fmt.Scan(&Option)

		Option = strings.ToLower(Option)

		switch Option {
		case ExitRegex.FindString(Option):
			os.Exit(0)
		case ClearRegex.FindString(Option):
			ClearScreen()
			PrintMenu()
		case "1":
			GetTWOM()
			PrintMenu()
		case "2":
			if !UseWASD {
				go WASDControls()
			} else {
				WASDChan <- struct{}{}
			}
			UseWASD = !UseWASD
			ClearScreen()
			PrintMenu()
		case "3":
			if !PReadMem {
				go ReadMem()
			} else {
				ReadMemChan <- struct{}{}
			}
			PReadMem = !PReadMem
			ClearScreen()
			PrintMenu()
		case "4":
			if !PFixCam {
				go FixCam()
			} else {
				FixCamChan <- struct{}{}
			}
			PFixCam = !PFixCam
			ClearScreen()
			PrintMenu()
		case "5":
			if !DisablePEffect {
				memory.NOP(HANDLE, Pencil, 9)
			} else {
				memory.WriteProcessMemory(HANDLE, Pencil, []byte{0xF3, 0x44, 0x0F, 0x10, 0x05, 0x35, 0x10, 0x4A, 0x00}, 9)
			}
			DisablePEffect = !DisablePEffect
			ClearScreen()
			PrintMenu()
		case "6":
			if !DisableRainEffect {
				memory.NOP(HANDLE, Rain, 7)
			} else {
				memory.WriteProcessMemory(HANDLE, Rain, []byte{0x0F, 0xB7, 0x05, 0xB8, 0xD7, 0x72, 0x00}, 7)
			}
			DisableRainEffect = !DisableRainEffect
			ClearScreen()
			PrintMenu()
		case "7":
			if !DisableOutlines {
				memory.NOP(HANDLE, Outline, 8)
			} else {
				memory.WriteProcessMemory(HANDLE, Outline, []byte{0xF3, 0x0F, 0x10, 0x0D, 0xD8, 0x2F, 0x60, 0x00}, 8)
			}
			DisableOutlines = !DisableOutlines
			ClearScreen()
			PrintMenu()
		case "8":
			if !ModifyWndProc {
				memory.NOP(HANDLE, WndProc, 5)
			} else {
				memory.WriteProcessMemory(HANDLE, WndProc, []byte{0x83, 0xE8, 0x02, 0x74, 0x26}, 5)
			}
			ModifyWndProc = !ModifyWndProc
			ClearScreen()
			PrintMenu()
		case "a":
			if !UseWASD {
				go WASDControls()
			}
			if !PReadMem {
				go ReadMem()
			}
			if !PFixCam {
				go FixCam()
			}
			if !DisablePEffect {
				memory.NOP(HANDLE, Pencil, 9)
			}
			if !DisableRainEffect {
				memory.NOP(HANDLE, Rain, 7)
			}
			if !DisableOutlines {
				memory.NOP(HANDLE, Outline, 8)
			}
			if !ModifyWndProc {
				memory.NOP(HANDLE, WndProc, 5)
			}

			UseWASD = true
			PReadMem = true
			PFixCam = true
			DisablePEffect = true
			DisableRainEffect = true
			DisableOutlines = true
			ModifyWndProc = true
			ClearScreen()
			PrintMenu()
		case "0":
			fmt.Print("Enter new Step Value: ")
			fmt.Scan(&Step)
			if Step <= 0 {
				Step = 0.7
			}
			ClearScreen()
			PrintMenu()
		case "i":
			Ceasefirep = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x008A7998, 0x4B0, 0x08, 0x10, 0x30, 0x00, 0x34)
			Intensityp = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x008A7998, 0x4B0, 0x08, 0x10, 0x30, 0x08, 0x34)
			WinterComesp = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x008A7998, 0x4B0, 0x08, 0x10, 0x30, 0x10, 0x34)
			WinterHarshnessp = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x008A7998, 0x4B0, 0x08, 0x10, 0x30, 0x18, 0x34)
			WinterLengthp = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x008A7998, 0x4B0, 0x08, 0x10, 0x30, 0x20, 0x34)

			rand.Seed(time.Now().UnixNano())
			Ceasefire = rand.Intn(13)
			Intensity = rand.Intn(3)
			WinterComes = 1 + rand.Intn(3)
			WinterHarshness = rand.Intn(3)
			WinterLength = rand.Intn(3)

			LocationsToChoose = 8 + Ceasefire

			memory.WriteProcessMemory(HANDLE, Ceasefirep, []byte{byte(Ceasefire)}, 1)
			memory.WriteProcessMemory(HANDLE, Intensityp, []byte{byte(Intensity)}, 1)
			memory.WriteProcessMemory(HANDLE, WinterComesp, []byte{byte(WinterComes)}, 1)
			memory.WriteProcessMemory(HANDLE, WinterHarshnessp, []byte{byte(WinterHarshness)}, 1)
			memory.WriteProcessMemory(HANDLE, WinterLengthp, []byte{byte(WinterLength)}, 1)
			PrintMenu()
		case "o":
			LocationsPTR = memory.Offsets(HANDLE, uintptr(BaseAddr), 0x008A7998, 0x800, 0x220, 0x00, 0x00, 0x120, 0x10)

			LocationsAddr = nil
			for k := range LocationsAdded {
				delete(LocationsAdded, k)
			}

			LocationsAddr = append(LocationsAddr, LocationsPTR)
			for i := 0; i < 31; i++ {
				LocationsAddr = append(LocationsAddr, LocationsAddr[len(LocationsAddr)-1]+56)
			}

			rand.Seed(time.Now().UnixNano())
			for i := 0; i < LocationsToChoose; i++ {
				id := rand.Intn(len(LocationsAddr))
				if LocationsAdded[id] != 0 {
					i--
				} else {
					LocationsAdded[id] = LocationsAddr[id]
				}
			}

			for id, pointer := range LocationsAdded {
				memory.WriteProcessMemory(HANDLE, pointer, []byte{byte(rand.Intn(LocationsData[id])), 0x00, 0x00, 0x00}, 4)
			}
			PrintMenu()
		case "p":
			ClearScreen()
			PrintPatches()
		case "q":
			fmt.Println("Attempting to Inject TWOMUHook")
			path, err := os.Getwd()
			if err != nil {
				fmt.Println("Failed to get Current Directory")
			}

			dllPath := fmt.Sprintf(`%v\TWOMUHook\x64\TWOMUHook.dll`, path)

			Valloc := memory.VirtualAllocEx(HANDLE, 0, uintptr(len(dllPath)+1), 0x00002000|0x00001000, 4)

			memory.WriteProcessMemory(HANDLE, Valloc, []byte(dllPath), uintptr(len(dllPath)+1))

			modKernel, err := syscall.LoadLibrary("kernel32.dll")
			if err != nil {
				memory.OutputDebugStringW(err.Error())
			}

			LoadLibrary, err := syscall.GetProcAddress(modKernel, "LoadLibraryA")
			if err != nil {
				memory.OutputDebugStringW(err.Error())
			}

			memory.CreateRemoteThread(HANDLE, 0, 0, LoadLibrary, Valloc, 0)
			PrintMenu()
		default:
			fmt.Println(Option, "is not a valid option.")
		}
	}
}

func WASDControls() {
	for {
		select {
		case <-WASDChan:
			return
		default:
			<-S.C
			Mutex.Lock()
			if pid := memory.GetWindowThreadProcessId(memory.GetForegroundWindow()); pid == TWOMPID {
				// W
				if memory.GetAsyncKeyState(0x57) {
					Y += Step
					memory.WriteProcessMemoryFloat(HANDLE, YPos, Y)
				}

				// A
				if memory.GetAsyncKeyState(0x41) {
					X -= Step
					memory.WriteProcessMemoryFloat(HANDLE, XPos, X)
				}

				// S
				if memory.GetAsyncKeyState(0x53) {
					Y -= Step
					memory.WriteProcessMemoryFloat(HANDLE, YPos, Y)
				}

				// D
				if memory.GetAsyncKeyState(0x44) {
					X += Step
					memory.WriteProcessMemoryFloat(HANDLE, XPos, X)
				}
			}
			Mutex.Unlock()
			if !S.Stop() {
				S.Reset(10 * time.Millisecond)
			}
		}
	}
}

func ReadMem() {
	for {
		select {
		case <-ReadMemChan:
			return
		default:
			<-S.C
			Mutex.Lock()

			XBuffer = memory.ReadProcessMemory(HANDLE, XPos, 8)
			X = math.Float32frombits(binary.LittleEndian.Uint32(XBuffer))

			YBuffer = memory.ReadProcessMemory(HANDLE, YPos, 8)
			Y = math.Float32frombits(binary.LittleEndian.Uint32(YBuffer))

			Mutex.Unlock()
			if !S.Stop() {
				S.Reset(10 * time.Millisecond)
			}
		}
	}
}

func FixCam() {
	for {
		select {
		case <-FixCamChan:
			return
		default:
			<-S.C
			CMBuffer = memory.ReadProcessMemory(HANDLE, CMode, 4)
			CM = binary.LittleEndian.Uint32(CMBuffer)

			if CM != 148602 {
				memory.WriteProcessMemory(HANDLE, CMode, []byte{0x7A, 0x44, 0x02}, 3)
			}
			if !S.Stop() {
				S.Reset(10 * time.Millisecond)
			}
		}
	}
}
